lib/logstash/filters/memcached.rb

```diff
@@ -67,28 +68,51 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
   # NOTE: in Memcached, a value of 0 (default) means "never expire"
   config :ttl, :validate => :number, :default => 0
 
+  # Tags the event on failure. This can be used in later analysis.
+  config :tag_on_failure, :validate => :array, :default => ["_memcached_failure"]
+
   public
 
   attr_reader :cache
 
   def register
-    if @ttl < 0
-      logger.error("ttl cannot be negative")
-      fail("invalid ttl: cannot be negative")
+    raise(LogStash::ConfigurationError, "'ttl' option cannot be negative") if @ttl < 0
+
+    @memcached_hosts = validate_connection_hosts
+    @memcached_options = validate_connection_options
+    begin
+      @cache = new_connection(@memcached_hosts, @memcached_options)
+    rescue => e
+      logger.error("failed to connect to memcached", hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      fail("failed to connect to memcached")
     end
-
-    @cache = establish_connection
-  end # def register
+    @connected = true
+  end
 
   def filter(event)
-    set_success = do_set(event)
-    get_success = do_get(event)
+    unless @connected
+      return unless reconnect(@memcached_hosts, @memcached_options)
+    end
 
-    filter_matched(event) if (set_success || get_success)
-  end # def filter
+    begin
+      set_success = do_set(event)
+      get_success = do_get(event)
+    rescue Dalli::NetworkError, Dalli::RingError => e
+      @tag_on_failure.each{|tag| event.tag(tag)}
+      logger.error("memcached communication error",  hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      close
+    else
```

Comment:
the `else` body is executed if not exception is raised so shouldn't more natural to move the `filter_matched(event) if (set_success || get_success)` into the `begin..rescue`?

Reply:
Yes, agree.

Reply:
pushed fix